### ISSUE TITLE
Need a way to add annotations without modifying capnp schema files

### ISSUE THREAD

**Issue by zenhack** [sandstorm-io/capnproto#323]
I'm currently working on trying to talk to the sandstorm API in Go. [capnpc-go][1] needs a few annotations to generate packages, and right now I'm having to do some munging of .capnp files from both sandstorm and those that ship with capnproto itself. It would be nice if I could just do something along the lines of:

    #include <util.capnp>
    using Go = import "/zombiezen.com/go/capnproto2/go.capnp";
    $Go.package("util");
    $Go.import("...");

Instead, I've currently got some goop in a makefile that copies the .capnp file and then appends stuff to it.

**Comment by zombiezen** [comment]
See also the discussion on #184.

**Comment by zenhack** [comment]
Another thing that was touched on in that thread and came up for me between opening this and now: needing to rename specific symbols requires an extra annotation; I hit this in the case of persistent and a couple other things. Having to add these in-line means we're basically maintaining a fork of the upstream schema, to get them to work with Go.

@kentonv, you mentioned in that thread that you'd thought about the problem and hadn't come up with anything that wasn't fairly complicated. Care to summarize what you've explored?

**Comment by kentonv** [comment]
At one point I was coming up with a hopelessly-over-engineered scheme in which you could have a separate file that imports the original and then applies transformations... but I don't totally remember the details of that.

We also faced this problem a lot in protobufs, with language-specific options as well as with `option optimize_for = LITE_RUNTIME`. We never came up with a great solution.

Maybe the simple solution is to add a way to declare an annotation on someone else's declaration, like:

    using Foo = import "foo.capnp";
    using MyLang = import "my-lang.capnp";

    annotate Foo $MyLang.package("blah");
    annotate Foo.Bar $MyLang.rename("Baz");

When the compiler sees `annotate` it would go and modify the schema for the symbol being annotated. If the same annotation is declared multiple times, and the declarations don't all match, then it's an error.

The weird thing about this is that now you have a side-effect depending on whether the compiler happened to parse this file during this execution. Depending on whether the modifier file was parsed, the output of generating `foo.capnp` could vary. That's really ugly, but I haven't thought of anything better. You also now need a way to tell the compiler to parse some extra files that don't happen to be imported anywhere, and that you don't actually want it to generate code for. Maybe:

    capnp compile -omylang foo.capnp --modifiers foo-modifiers.capnp

**Comment by paulproteus** [comment]
For the Sandstorm capnp files, at least, is the above scheme better than us accepting pull requests with language-specific stuff?

**Comment by kentonv** [comment]
@paulproteus Yes because in order to add the annotations directly to the original files, we would also need to add dependencies to all those files onto all of the different languages' annotation definitions, which means Sandstorm would now depend on every language's capnp implementation.

**Comment by zombiezen** [comment]
FWIW, for my personal hackery "monorepo", I vendor in capnproto and go-capnproto2 and modify capnproto's schemas in-place to add the Go annotations.  This way, there's only one copy of the schemas.  While it's a tad annoying to initially set up, it's conceptually simple.

My vote is for punting.  I'm no longer convinced it's an issue that should be solved.

**Comment by zenhack** [comment]
Something similar to the external annotation thing @kentonv suggested had gone through my head, but I don't think this is a severe enough problem to warrant implementing an unsatisfying solution; if we can think of something that feels like "the right thing" then I'm for it, but right now the manual tweaking we've done for the Go stuff feels like a better approach than anything else that's been proposed.

I'll also point out that some of our specific use cases could be solved by the compiler plugin being a little smarter; we could use algorithms that are guaranteed to generate valid identifiers with no collisions, for example. This would be enough to get rid of the manual tweaks I've had to do, and reduce it to just the bits that were easy enough to automate. It wouldn't solve stuff like the optimization hints, but that's not a problem for the Go implementation at least.

I'm for leaving this open for now; it would be nice to have a solution, and I don't think it's an insurmountable problem, but it's going to require some thought, and there's clearly no rush.

**Comment by zombiezen** [comment]
+1, capnpc-go should definitely be smarter about things.  I'll open up a couple of issues just to get my current thinking out there.

**Comment by chaoflow** [comment]
Would annotations on using directives (like discussed here #411 ) be a solution for this?

**Comment by zenhack** [comment]
Maybe. Something like:

```
using Go = import "/std/go.capnp";
using websession = import "web-session.capnp" $Go.package(name = "websession", path = "zenhack.net/go/sandstorm/capnp/websession");
```

How does this compare to the separate-file annotation solution that @kentonv mentioned?

**Comment by kentonv** [comment]
Hmm, I don't see how that would work, since when generating code for `web-session.capnp` itself, you presumably need to know what package it's in. But with this solution, its package is only known to files that import it. (Also what happens if different files disagree?)

**Comment by sashka** [comment]
In my company's internal fork we're experimenting with syntax like this.
It's in alfa stage yet, but already works in our go environment.

```
# /capnproto/go/aircraft.capnp
# run capnp compile -ogo:/capnproto/go/ *.capnp

@0xb8e783ba0d624ec7;

// Request schema file right from the current code generator plugin.
using Go = import "@go.capnp";
using A = import "../annotations.capnp";
// Copy whole schema to the scope of this file.
// Treat this like the whole original file (excluding @id) is copy-pasted here.
embed "../aircraft.capnp";

// Annotations on a file level work as expected.
$Go.package("aircraft");
$Go.import("mokote.com/capnproto/reflect")
$Go.generate_json_serializer;
$Go.generate_binary_serializer;

// Keyword "annotate" lets to add an annotation to anything I can reference by a name.
annotate Aircraft.companyId $A.required;
```

Known problem, not yet solved: no way to annotate anonymous groups or unions.

**Comment by sashka** [comment]
In my company's internal fork we're experimenting with syntax like this.
It's in alfa stage yet, but already works in our go environment.

```
# /capnproto/go/aircraft.capnp
# run capnp compile -ogo:/capnproto/go/ *.capnp

@0xb8e783ba0d624ec7;

// Request schema file right from the current code generator plugin.
using Go = import "@go.capnp";
using A = import "../annotations.capnp";

// Copy whole schema to the scope of this file.
// Treat this like the whole original file (excluding @id) is pasted here.
// It's pasted with all its imports interpreted within original directory.
embed "../aircraft.capnp";

// Annotations on a file level work as expected.
$Go.package("aircraft");
$Go.generate_json_serializer;
$Go.generate_binary_serializer;

// Keyword "annotate" lets to add an annotation to anything I can reference by a name.
// It works only in a file scope.
annotate Aircraft.companyId $A.required;
```

Known problem, not yet solved: there's no way to annotate anonymous groups or unions.

**Comment by zenhack** [comment]
Spitballing, still not in love with this, but: what if we had a way to mark an annotation/import as "it's okay if you can't find this?" This wouldn't solve all of the problems we're having, but it would make @paulproteus's suggestion more palatable, in that we could upstream the namespacing of things without needing to merge in e.g. go.capnp itself.

I'm digging back into my haskell implementation, and I think there I've come up with a way to make everything work without having any language-specific annotations, but it relys on some properties of the module system that won't work in every language.

**Comment by sashka** [comment]
We have this mark in our fork, but I’m not yet happy with this. 

Our python, go and js plugins carry their schemas inside and they know how to respond `import “@schemaname.capnp”` or even `import "@id”`. We may mix language annotations in one proxy schema with no errors emitted when current plugin doesn’t know the name or id. 

I still think this approach is a dirty hack, that’s why we’re still experimenting with this and another ways. 

@kentonv should I send you our internal Enhancement Proposals on features we implemented internally?

**Comment by kentonv** [comment]
Hi,

Sorry for the long delay in replying to this. It's one of those "Hmm, I need to think carefully about this and I don't have time right now, maybe tomorrow..." and then... it sits in my inbox for a month. :(

I don't quite understand how this works. If these imports are processed by the plugin, how does the compiler parse them, since parsing happens before the plugin is invoked? Does it somehow communicate to the plugin the names of the annotations rather than their IDs?

Maybe the easiest way for me to understand this would be if I could see the changes to schema.capnp that you made. Is that visible somewhere?

**Comment by zombiezen** [comment]
Go Cap'n Proto does this here: https://github.com/capnproto/go-capnproto2/tree/master/std/capnp
